src/Microsoft.Diagnostics.Monitoring.EventPipe/EventPipeStreamProvider.cs

```diff
@@ -28,7 +30,44 @@ public async Task<Stream> ProcessEvents(DiagnosticsClient client, TimeSpan durat
             EventPipeSession session = null;
             try
             {
-                session = await client.StartEventPipeSessionAsync(_sourceConfig.GetProviders(), _sourceConfig.RequestRundown, _sourceConfig.BufferSizeInMB, cancellationToken).ConfigureAwait(false);
+                IEnumerable<EventPipeProvider> providers = _sourceConfig.GetProviders();
+                int bufferSizeInMB = _sourceConfig.BufferSizeInMB;
+                long rundownKeyword = _sourceConfig.RundownKeyword;
+                RetryStrategy retryStrategy = _sourceConfig.RetryStrategy;
+                bool retry = true;
+                while (retry)
+                {
+                    retry = false;
+                    try
+                    {
+                        EventPipeSessionConfiguration config = new(providers, bufferSizeInMB, (rundownKeyword != 0), true, rundownKeyword);
+                        session = await client.StartEventPipeSessionAsync(config, cancellationToken).ConfigureAwait(false);
+                    }
+                    catch (UnsupportedCommandException e)
+                    {
+                        if (retryStrategy == RetryStrategy.DropKeywordKeepRundown)
+                        {
+                            Debug.Assert(rundownKeyword != 0);
+                            Debug.Assert(rundownKeyword != EventPipeSession.DefaultRundownKeyword);
+                            retry = true;
+                            retryStrategy = RetryStrategy.DoNotRetry;
+                            rundownKeyword = 0;
+                        }
+                        else if (retryStrategy == RetryStrategy.DropKeywordDropRundown)
+                        {
+                            Debug.Assert(rundownKeyword != 0);
+                            Debug.Assert(rundownKeyword != EventPipeSession.DefaultRundownKeyword);
+                            retry = true;
+                            retryStrategy = RetryStrategy.DoNotRetry;
+                            rundownKeyword = 0;
+                        }
+                        else
+                        {
+                            Debug.Assert((rundownKeyword == 0) || (rundownKeyword == EventPipeSession.DefaultRundownKeyword));
```

Comment:
Is seems like it should be legal for someone to create a Profile where RundownKeyword is non-zero non-default and the retry strategy is DoNotRetry. In that case this assert would trigger unnecessarily.

Reply:
Agreed